pycbc/waveform/generator.py

```diff
@@ -0,0 +1,327 @@
+
+        # if detectors are provided, convert to detector type; also ensure that
+        # location variables are specified
+        if detectors is not None:
+            self.detectors = [Detector(det) for det in detectors]
+            missing_args = [arg for arg in self.location_args if not
+                (arg in self.current_params or arg in self.variable_args)]
+            if any(missing_args):
+                raise ValueError("detectors provided, but missing location "
+                    "parameters %s. " %(', '.join(missing_args)) +
+                    "These must be either in the frozen params or the "
+                    "variable args.")
+        else:
+            self.detectors = [None]
+
+    @property
+    def detector_names(self):
+        if self.detectors == [None]:
+            return [None]
+        else:
+            return [det.name for det in self.detectors]
+
+    def generate(self, *args):
+        """Generates a waveform, applies a time shift and the detector response
+        function."""
+        self.current_params.update(dict(zip(self.variable_args, args)))
+        rfparams = {param: self.current_params[param]
+            for param in self.rframe_generator.variable_args}
+        hp, hc = self.rframe_generator.generate_from_kwargs(**rfparams)
+        if self.epoch is not None:
+            hp._epoch = self.epoch
+            hc._epoch = self.epoch
+        # we'll need the sample points for applying the time shift
+        if self.sample_points is None:
+            fseries = hp.sample_frequencies.numpy()
+        else:
+            fseries = self.sample_points
+            # in case the waveform terminated prior to the largest frequecy
+            # in the sample points
+            if len(fseries) > len(hp):
+                fseries = fseries[:len(hp)]
+            # in case the waveform terminated at a frequency larger than the
+            # largest desired in the sample points
+            if len(fseries) < len(hp):
+                hp = hp[:len(fseries)]
+                hc = hc[:len(fseries)]
+        h = {}
+        if self.detectors != [None]:
+            for det in self.detectors:
+                tc = self.current_params['tc'] + \
+                    det.time_delay_from_earth_center(self.current_params['ra'],
+                                                     self.current_params['dec'],
+                                                     self.current_params['tc'])
+                # apply the time shift
+                thishp = apply_fd_time_shift(hp, tc,
+                        fseries=fseries, copy=True)
+                thishc = apply_fd_time_shift(hc, tc,
+                        fseries=fseries, copy=True)
+                # apply detector response function
+                fp, fc = det.antenna_pattern(self.current_params['ra'],
+                            self.current_params['dec'],
+                            self.current_params['polarization'],
```

Comment:
This section can be simplified by combining the apply_fd_time_shifts into a single call at the end, and doing the antenna weighting before the time shift.